paddle/fluid/framework/tensor.h

```diff
@@ -37,23 +37,84 @@ class Tensor {
+    auto md = mkldnn::memory::desc({ltz}, data_type, format);
+    auto& pool = platform::DeviceContextPool::Instance();
+    auto place = paddle::platform::CPUPlace();
+    auto* dev_ctx =
+        dynamic_cast<platform::MKLDNNDeviceContext*>(pool.Get(place));
+    auto& cpu_engine = dev_ctx->GetEngine();
+    mem_pd_ = mkldnn::memory::primitive_desc(md, cpu_engine);
+  }
+
+  inline mkldnn::memory::primitive_desc MKLDNNMemPrimDesc() const {
+    return mem_pd_;
+  }
+
+  inline void setMKLDNNMemPrimDesc(mkldnn::memory::primitive_desc& mem_pd) {
+    // Internally MKL-DNN is just copying (increasing reference counter)
+    // to shared_ptr. So asignment should be quite cheap
+    mem_pd_ = mem_pd;
+  }
 
-  inline void set_format(const mkldnn::memory::format format) {
-    format_ = format;
+  void CreateMemoryPrimDescFromDims(
+      mkldnn::memory::format fmt,
+      mkldnn::memory::data_type data_type = mkldnn::memory::data_type::f32) {
+    mkldnn_memory_desc_t mem_fmt;
+
+    mem_fmt.primitive_kind = mkldnn_memory;
+    std::vector<int> ltz = paddle::framework::vectorize2int(dims());
+    mem_fmt.ndims = ltz.size();
+    for (unsigned int i = 0; i < ltz.size(); ++i) {
+      mem_fmt.dims[i] = ltz[i];  // logical dimensions (nchw format,
+                                 // regardless physical layout)
+    }
+    mem_fmt.data_type = static_cast<mkldnn_data_type_t>(data_type);
+    mem_fmt.format = static_cast<mkldnn_memory_format_t>(fmt);
+
+    unsigned int total_stride = 1;
+    for (int i = ltz.size() - 1; i >= 0; --i) {
+      mem_fmt.layout_desc.blocking.padding_dims[i] =
+          ltz[i];  // logical dimensions (nchw format, regardless physical
+                   // layout)
+      mem_fmt.layout_desc.blocking.block_dims[i] = 1;
+      mem_fmt.layout_desc.blocking.offset_padding_to_data[i] = 0;  // no offset
+      mem_fmt.layout_desc.blocking.strides[0][i] = total_stride;
+      mem_fmt.layout_desc.blocking.strides[1][i] = 1;
+      total_stride *= ltz[i];
+    }
+    mem_fmt.layout_desc.blocking.offset_padding = 0;  // no initial offset
+
+    auto& pool = platform::DeviceContextPool::Instance();
+    auto place = paddle::platform::CPUPlace();
+    auto* dev_ctx =
+        dynamic_cast<platform::MKLDNNDeviceContext*>(pool.Get(place));
+    auto& cpu_engine = dev_ctx->GetEngine();
+    mem_pd_ = mkldnn::memory::primitive_desc(mem_fmt, cpu_engine);
   }
 
  protected:
   /**
    * @brief the detail format of memory block which have layout as kMKLDNN
    *
    * @note MKLDNN lib support various memory format like nchw, nhwc, nChw8C,
-   *       nChw16c, etc. For a MKLDNN memory block, layout will be set as
-   *       DataLayout::kMKLDNN meanwhile detail memory format will be kept in
-   *       this field.
+   *       nChw16c, etc. For a MKLDNN memory block, we store memory descriptor
    */
-
-  mkldnn::memory::format format_ = mkldnn::memory::format::format_undef;
+  mutable mkldnn::memory::primitive_desc mem_pd_;
```

Comment:
what does this contain? Can it be mapped to current tensor apis? such as numel,layer?

Reply:
@panyx0718 mem_pd_ is a primitive descriptor of MKL-DNN eg. This a structure that describes the implementation details of memory primitive eg. object does represent memory. So in that case mem_pd_ describes the dims of memory allocations , sizes, layout(way how memory is arranged, strides), and location of memory eg. cpu (MKL-DNN currently only supports CPU).